### PR TITLE
datastore.py: update to meet new datastore builtin PR.

### PR DESCRIPTION
**Pull request by rustyrussell**
The new datastore PR now has a generation count; this passes
the tests against that now.

Also copies tests from c-lightning, and adds ugprade test.

Changelog-None

### CODE REVIEW

File: datastore/datastore-plugin.py [hunk 1]
```diff
@@ -2,90 +2,199 @@
...
-    if key in plugin.datastore:
-        return {'datastore': [datastore_entry(key, plugin.datastore[key])]}
-    return {'datastore': []}
+
+    key = normalize_key(key)
+    ret = []
+    prev = None
+    for khex, e in sorted(plugin.datastore.items()):
+        k = hex_to_key(khex)
+        print("... {}".format(k))
```
Comment (cdecker) [inline datastore/datastore-plugin.py:162]: Likely to be noisy and not helpful (we're printing hex-encoded strings)

**Review by cdecker — COMMENTED** [review]
Some minor comments, and not sure that going with hex-encoded keys is the right thing to do: either they are string, and we should keep them as such, or they aren't, in which case we can store them in the shelve directly.

File: datastore/datastore-plugin.py [hunk 2]
```diff
@@ -2,90 +2,199 @@
...
-def datastore_entry(key, data):
+# A singleton to most commands turns into a [].
+def normalize_key(key: Union[Sequence[str], str]) -> List[str]:
+    if not isinstance(key, list) and not isinstance(key, tuple):
+        key = [key]
+    return key
+
+# We turn list into nul-separated hexbytes for storage
+def key_to_hex(key: Sequence[str]) -> str:
+    return b'\0'.join([bytes(k, encoding='utf8') for k in key]).hex()
```
[the remark this answers is not in the document]
Comment (rustyrussell) [reply datastore/datastore-plugin.py:32]: Because shelve requires all keys be strings :(

[the remark this answers is not in the document]
Comment (cdecker) [reply datastore/datastore-plugin.py:33]: Oh, didn't know that :-(

**Comment by cdecker** [comment]
ACK https://github.com/lightningd/plugins/pull/283/commits/a0be5d26411c897b1e013d4655ff2a777f611b86

**Pull request merged**